File: SentinelAI-v2-main/sentinelai-v2/backend/agents/url_agent.py
"""
SentinelAI v2.0 — URL + Intel Agent
Model: SmolLM2-1.7B-Instruct via Ollama
Tasks: URL entropy, homoglyph detection, typosquatting, TLD trust, threat feed queries
"""

import re
import math

import os
import httpx
from typing import Optional

MODEL_NAME = "smollm2:1.7b"

SUSPICIOUS_TLDS = {
    '.tk', '.ml', '.ga', '.cf', '.gq', '.xyz', '.top', '.club',
    '.work', '.date', '.racing', '.win', '.bid', '.stream', '.click',
    '.link', '.loan', '.trade', '.cricket', '.science', '.party'
}

HOMOGLYPHS = {
    'а': 'a', 'е': 'e', 'о': 'o', 'р': 'p', 'с': 'c', 'у': 'y',
    'х': 'x', 'ѕ': 's', 'і': 'i', 'ј': 'j', 'ɡ': 'g', 'ɩ': 'l',
    '0': 'o', '1': 'l'
}

BRAND_DOMAINS = [
    'google.com', 'facebook.com', 'apple.com', 'microsoft.com',
    'amazon.com', 'paypal.com', 'netflix.com', 'instagram.com',
    'twitter.com', 'linkedin.com', 'chase.com', 'bankofamerica.com',
    'wellsfargo.com', 'dropbox.com', 'outlook.com', 'icloud.com'
]


def _url_entropy(url: str) -> float:
    """Calculate Shannon entropy of a URL string."""
    if not url:
        return 0.0
    freq = {}
    for c in url:
        freq[c] = freq.get(c, 0) + 1
    length = len(url)
    return -sum((count / length) * math.log2(count / length) for count in freq.values())


def _count_homoglyphs(hostname: str) -> int:
    return sum(1 for c in hostname if c in HOMOGLYPHS)


def _check_brand_impersonation(hostname: str) -> Optional[str]:
    for brand in BRAND_DOMAINS:
        brand_name = brand.split('.')[0]
        if brand_name in hostname and not hostname.endswith(brand):
            return brand
    return None


def analyze_heuristic(url: str) -> dict:
    """Run heuristic URL analysis (no LLM needed)."""
    threats = []
    score = 0

    try:
        from urllib.parse import urlparse
        parsed = urlparse(url)
        hostname = (parsed.hostname or "").lower()

        # TLD check
        tld = '.' + hostname.split('.')[-1] if '.' in hostname else ''
        if tld in SUSPICIOUS_TLDS:
            score += 20
            threats.append({"type": "suspicious-tld", "detail": f"TLD {tld} is commonly used in phishing"})

        # Homoglyphs
        hg_count = _count_homoglyphs(hostname)
        if hg_count > 0:
            score += 30
            threats.append({"type": "homoglyph", "detail": f"{hg_count} homoglyph character(s) in hostname"})

        # Brand impersonation
        brand = _check_brand_impersonation(hostname)
        if brand:
            score += 35
            threats.append({"type": "brand-impersonation", "detail": f"Possible impersonation of {brand}"})

        # IP hostname
        if re.match(r'^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$', hostname):
            score += 15
            threats.append({"type": "ip-hostname", "detail": "URL uses IP address instead of domain"})

        # Subdomain count
        subdomain_count = hostname.count('.') - 1
        if subdomain_count > 3:
            score += 10
            threats.append({"type": "excessive-subdomains", "detail": f"{subdomain_count} subdomains"})

        # URL entropy
        entropy = _url_entropy(url)
        if entropy > 4.5:
            score += 10
            threats.append({"type": "high-entropy", "detail": f"URL entropy {entropy:.2f} (suspicious)"})

        # HTTP
        if parsed.scheme == 'http':
            score += 10
            threats.append({"type": "no-https", "detail": "Connection is not encrypted"})

        # URL length
        if len(url) > 200:
            score += 5
            threats.append({"type": "long-url", "detail": f"URL length {len(url)} chars"})

    except Exception:
        score += 10
        threats.append({"type": "parse-error", "detail": "URL could not be parsed"})

    return {
        "agent": "url-agent",
        "model": MODEL_NAME,
        "score": min(score, 100),
        "threats": threats
    }


# Removed _check_threat_apis because we now use fully offline threat caching


async def analyze_llm(url: str) -> dict:
    """Enhanced analysis using SmolLM2-1.7B-Instruct via Ollama and Threat APIs."""
    heuristic = analyze_heuristic(url)
    
    # Step 07 Architecture: Threat API Layer Lookups (OFFLINE)
    from backend.threat_cache import lookup_threat
    local_threat = lookup_threat(url)
    if local_threat:
        heuristic["threats"].append({"type": local_threat["type"], "detail": f"Matched in offline {local_threat['source']} feed"})
        heuristic["score"] = max(heuristic["score"], local_threat["risk"])

    prompt = f"""You are a cybersecurity URL analysis agent. Analyze this URL for phishing/malware indicators.
URL: {url}

Heuristic findings: {heuristic['threats']}

Provide a JSON response with:
- "risk_assessment": brief risk summary
- "additional_threats": any threats the heuristics may have missed
- "confidence": 0-100

Respond ONLY with valid JSON."""

    try:
        from backend.ollama_engine import generate_async, is_available
        if is_available():
            llm_response = await generate_async("url-agent", prompt)
            heuristic["llm_analysis"] = llm_response
        else:
            heuristic["llm_analysis"] = "AirLLM not installed"
    except Exception as e:
        heuristic["llm_analysis"] = f"LLM unavailable: {str(e)}"

    return heuristic

File: SentinelAI-v2-main/sentinelai-v2/agents/url-agent.js
/**
 * SentinelAI v2.0 — URL Agent
 * URL & Intel analysis — pattern matching, homoglyph detection, TLD reputation.
 * Stub LLM call for advanced analysis (points to Ollama).
 */

const SUSPICIOUS_TLDS = [
  '.tk', '.ml', '.ga', '.cf', '.gq', '.xyz', '.top', '.club',
  '.work', '.date', '.racing', '.win', '.bid', '.stream', '.click',
  '.link', '.loan', '.trade', '.cricket', '.science', '.party'
];

const HOMOGLYPHS = {
  'а': 'a', 'е': 'e', 'о': 'o', 'р': 'p', 'с': 'c', 'у': 'y',
  'х': 'x', 'ѕ': 's', 'і': 'i', 'ј': 'j', 'ɡ': 'g', 'ɩ': 'l'
};

const BRAND_DOMAINS = [
  'google.com', 'facebook.com', 'apple.com', 'microsoft.com',
  'amazon.com', 'paypal.com', 'netflix.com', 'instagram.com',
  'twitter.com', 'linkedin.com', 'chase.com', 'bankofamerica.com',
  'wellsfargo.com', 'dropbox.com', 'outlook.com', 'icloud.com'
];

class URLAgent {
  constructor() {
    this.name = 'url-agent';
  }

  /**
   * Analyze a URL for threat signals
   * @param {string} url - The URL to analyze
   * @returns {Object} - { score: 0-100, threats: [], details: {} }
   */
  analyze(url) {
    const threats = [];
    let score = 0;

    try {
      const parsed = new URL(url);
      const hostname = parsed.hostname.toLowerCase();
      const fullUrl = parsed.href.toLowerCase();

      // 1. Check suspicious TLD
      const tld = '.' + hostname.split('.').pop();
      if (SUSPICIOUS_TLDS.includes(tld)) {
        score += 20;
        threats.push({ type: 'suspicious-tld', detail: `TLD ${tld} is commonly used in phishing` });
      }

      // 2. Check for homoglyphs in hostname
      const homoglyphCount = this._countHomoglyphs(hostname);
      if (homoglyphCount > 0) {
        score += 30;
        threats.push({ type: 'homoglyph', detail: `${homoglyphCount} homoglyph character(s) detected in hostname` });
      }

      // 3. Check for brand impersonation
      const brandMatch = this._checkBrandImpersonation(hostname);
      if (brandMatch) {
        score += 35;
        threats.push({ type: 'brand-impersonation', detail: `Possible impersonation of ${brandMatch}` });
      }

      // 4. Check for IP address hostname
      if (/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/.test(hostname)) {
        score += 15;
        threats.push({ type: 'ip-hostname', detail: 'URL uses IP address instead of domain name' });
      }

      // 5. Check for excessive subdomains
      const subdomainCount = hostname.split('.').length - 2;
      if (subdomainCount > 3) {
        score += 10;
        threats.push({ type: 'excessive-subdomains', detail: `${subdomainCount} subdomains detected` });
      }

      // 6. Check for suspicious URL patterns
      if (/login|signin|verify|secure|account|update|confirm/i.test(fullUrl) && SUSPICIOUS_TLDS.includes(tld)) {
        score += 15;
        threats.push({ type: 'phishing-pattern', detail: 'Login/verify keywords with suspicious TLD' });
      }

      // 7. Check for data URI
      if (parsed.protocol === 'data:') {
        score += 25;
        threats.push({ type: 'data-uri', detail: 'Data URI can hide malicious content' });
      }

      // 8. Check for non-standard port
      if (parsed.port && !['80', '443', ''].includes(parsed.port)) {
        score += 5;
        threats.push({ type: 'non-standard-port', detail: `Non-standard port: ${parsed.port}` });
      }

      // 9. Check URL length
      if (fullUrl.length > 200) {
        score += 5;
        threats.push({ type: 'long-url', detail: `Unusually long URL (${fullUrl.length} chars)` });
      }

      // 10. HTTP instead of HTTPS
      if (parsed.protocol === 'http:') {
        score += 10;
        threats.push({ type: 'no-https', detail: 'Connection is not encrypted (HTTP)' });
      }

    } catch(e) {
      score += 10;
      threats.push({ type: 'invalid-url', detail: 'URL could not be parsed' });
    }

    return {
      agent: this.name,
      score: Math.min(score, 100),
      threats,
      timestamp: Date.now()
    };
  }

  _countHomoglyphs(hostname) {
    let count = 0;
    for (const char of hostname) {
      if (HOMOGLYPHS[char]) count++;
    }
    return count;
  }

  _checkBrandImpersonation(hostname) {
    for (const brand of BRAND_DOMAINS) {
      const brandName = brand.split('.')[0];
      // Check if hostname contains brand name but isn't the actual domain
      if (hostname.includes(brandName) && !hostname.endsWith(brand)) {
        return brand;
      }
    }
    return null;
  }

  /**
   * Stub LLM call — would call Ollama phi4-mini for advanced URL analysis
   */
  async analyzeLLM(url) {
    // Stub: return heuristic result
    return this.analyze(url);
  }
}

// Export
if (typeof globalThis !== 'undefined') {
  globalThis.SentinelURLAgent = URLAgent;
}
